### ISSUE TITLE
Adding React-Toolbox to the feature/postcss-sass branch.

### ISSUE THREAD

**Issue by coreygo** [ctrlplusb/react-universally#323]
Looking to integrate the react-toolbox for Material Design UI elements.
https://github.com/react-toolbox/react-toolbox

There seems to be some issues with the webpack/postcss/sass-loader processing. For example the below after using `import { Button } from 'react-toolbox/lib/button';`

```bash
==> SERVER -> Error in server execution, check the console for more info.
/Users/corey/Projects/coreygo/coreygo/node_modules/react-toolbox/lib/ripple/theme.scss:1
(function (exports, require, module, __filename, __dirname) { @import "../colors";
                                                              ^
SyntaxError: Invalid or unexpected token
    at Object.exports.runInThisContext (vm.js:76:16)
    at Module._compile (module.js:542:28)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/corey/Projects/coreygo/coreygo/node_modules/react-toolbox/lib/ripple/index.js:13:14)
    at Module._compile (module.js:570:32)
```

**Comment by strues** [comment]
I've never been able to get react-toolbox to work with any universal webpack configurations. That being said, you'll need to add configuration to sass-loader.

```
  sassLoader: {
    data: '@import "theme/_config.scss";',  // the sass file where you import react toolbox
    includePaths: __Shared_Folder_Path__ 
  },
```

**Comment by coreygo** [comment]
Seems they've switched from SASS to PostCSS, using "react-toolbox": "^2.0.0-beta.4", still seems to have issues with or without the sass-loader.

Was able to basically rewrite the components and get them to work but that's a rather dumb workaround :)

**Comment by strues** [comment]
I'll see about digging into this some more this afternoon. Its always something I'd like to use, but never actually do because of this very issue.

**Comment by coreygo** [comment]
Cool. The readme.md for "react-toolbox": "^2.0.0-beta.4" isn't up to date :)

Here's what my fork of react-universally currently includes, sort of combining flow, postcss-sass, and apollo feature branches, also added bluebird and redux-saga:

`package.json`

```json
  "dependencies": {
    "apollo-client": "^0.7.1",
    "apollo-server": "^0.3.3",
    "app-root-dir": "1.0.2",
    "bluebird": "^3.4.7",
    "body-parser": "^1.15.2",
    "code-split-component": "2.0.0-alpha.5",
    "colors": "1.1.2",
    "compression": "1.6.2",
    "dotenv": "4.0.0",
    "express": "4.14.0",
    "graphql": "^0.8.2",
    "graphql-tag": "^1.2.2",
    "graphql-tools": "^0.9.0",
    "helmet": "3.3.0",
    "hpp": "0.2.1",
    "normalize.css": "5.0.0",
    "offline-plugin": "4.5.4",
    "react": "15.4.2",
    "react-addons-css-transition-group": "^15.4.2",
    "react-apollo": "^0.8.1",
    "react-dom": "15.4.2",
    "react-helmet": "3.3.0",
    "react-jobs": "^0.5.0",
    "react-redux": "^5.0.2",
    "react-router": "4.0.0-alpha.6",
    "react-toolbox": "^2.0.0-beta.4",
    "redux": "^3.6.0",
    "redux-saga": "^0.14.2",
    "serialize-javascript": "1.3.0",
    "user-home": "2.0.0",
    "uuid": "3.0.1"
  },
  "devDependencies": {
    "assets-webpack-plugin": "3.5.0",
    "babel-cli": "6.18.0",
    "babel-core": "6.21.0",
    "babel-eslint": "7.1.1",
    "babel-jest": "18.0.0",
    "babel-loader": "6.2.10",
    "babel-plugin-syntax-flow": "^6.18.0",
    "babel-plugin-transform-flow-strip-types": "^6.21.0",
    "babel-plugin-transform-react-jsx-self": "6.11.0",
    "babel-plugin-transform-react-jsx-source": "6.9.0",
    "babel-polyfill": "6.20.0",
    "babel-preset-env": "1.1.8",
    "babel-preset-latest": "6.16.0",
    "babel-preset-react": "6.16.0",
    "babel-preset-stage-3": "6.17.0",
    "babel-template": "6.16.0",
    "chokidar": "1.6.1",
    "cross-env": "3.1.4",
    "css-loader": "0.26.1",
    "enzyme": "2.7.0",
    "enzyme-to-json": "1.4.5",
    "eslint": "3.13.1",
    "eslint-config-airbnb": "14.0.0",
    "eslint-plugin-flowtype": "2.29.2",
    "eslint-plugin-import": "2.2.0",
    "eslint-plugin-jsx-a11y": "3.0.2",
    "eslint-plugin-react": "6.9.0",
    "extract-text-webpack-plugin": "^2.0.0-beta.5",
    "file-loader": "0.9.0",
    "flow-bin": "^0.37.4",
    "flow-coverage-report": "^0.2.0",
    "flow-typed": "^2.0.0",
    "glob": "7.1.1",
    "happypack": "3.0.2",
    "html-webpack-plugin": "2.26.0",
    "isomorphic-fetch": "^2.2.1",
    "jest": "18.1.0",
    "md5": "2.2.1",
    "node-notifier": "4.6.1",
    "node-sass": "^4.3.0",
    "postcss": "^5.2.10",
    "postcss-cssnext": "^2.9.0",
    "postcss-import": "^9.1.0",
    "postcss-loader": "^1.2.2",
    "postcss-simple-vars": "^3.0.0",
    "react-addons-test-utils": "15.4.2",
    "react-hot-loader": "3.0.0-beta.6",
    "regenerator-runtime": "0.10.1",
    "rimraf": "2.5.4",
    "sass-loader": "^4.1.1",
    "semver": "5.3.0",
    "source-map-support": "0.4.8",
    "style-loader": "0.13.1",
    "webpack": "2.2.0-rc.3",
    "webpack-bundle-analyzer": "2.2.1",
    "webpack-dev-middleware": "1.9.0",
    "webpack-hot-middleware": "2.15.0",
    "webpack-md5-hash": "0.0.5",
    "webpack-node-externals": "1.5.4"
  }
```

**Comment by strues** [comment]
@coreygo I would look into this https://github.com/react-toolbox/react-toolbox-themr 

I'll be trying it later this afternoon, but it should solve the problem.

**Comment by RishabhJain96** [comment]
Was there a solution to this? Other libraries aside from react-toolbox also run into this issue.

**Comment by diondirza** [comment]
@RishabhJain96 please look into my setup in my branch of **[feature/react-toolbox-sass](https://github.com/diondirza/react-universally/tree/feature/react-toolbox-sass)**, I've been able to make react-toolbox@1.3.4 running without problem so far.

I will make a PR for this branch anyway. Cheers.

**Comment by coreygo** [comment]
@diondirza your diff worked against my fork. So react-toolbox@1.3.4 is working well now. I was hoping to get it sorted with react-toolbox@2.0.0-beta.6 but for now it probably makes more sense to stick with the former version.

Thanks for sorting that.

**Comment by coreygo** [comment]
@diondirza could you clarify something for me? [toolbox.theme.scss](https://github.com/ctrlplusb/react-universally/compare/master...diondirza:feature/react-toolbox-sass#diff-10b9834f1669f4bb6947d5ca5cc7a346) seems to have no effect over the input color styling. Am I missing something about how it is being used? Clearing the file makes no difference.

**Comment by diondirza** [comment]
@coreygo If you look at their documentation there are two ways to theming react-toolbox. By using sass loader configuration in `data` prop to override some of their global variables or specifically per component you want. Like the way I put it on Home.js

configFactory.js line 410 & 417
```
{
  path: 'sass-loader',
  options: {
    sourceMap: true,
    sourceMapContents: true,
    outputStyle: 'expanded',
    data: `@import "${path.resolve(appRootDir.get(), './src/shared/components/DemoApp/toolbox.theme.scss')}";`,
  },
},
```

In that branch I used both of them so someone could pick what way they want. There is some glitch while you theming globally using file `toolbox.theme.scss`, if you run your app in development env it won't be applied, yet if in production env it will run properly. There must be something not right with how inline css loaded with webpack. Yet I don't know until now how to fix it.

**Comment by ctrlplusb** [comment]
Closing this for now.

**Comment by unleashit** [comment]
Anyone ever get react-toolbox ver2 (with postcss instead of sass) to work with server rendering? It's rendering alright on the client but postcss seems not to run on the server for some reason. No matter what I do, I keep getting the unexpected token : when it tries to process css vars. At one point I had it loading fine in the browser (but still getting the server error), but now I just get a white page with ERR connection refused as soon as I import anything from RT. I'm using the same config as before, so I guess a package update must be responsible for the change in behavior. Either way, something's not right. It's strange, because postcss (at least autoprefixer) and css modules are running without issue for my custom SASS.

Anyhoo, in case anyone has any insight here's my config. Would love to get this running without having to resort to the themr tool. Note that I've made an explicit config for RT, and have explicitly excluded it from the general css configs.

```
        ifElse(isClient || isServer)(
          mergeDeep(
            {
              test: /\.css$/,
              include: path.join(appRootDir.get(), 'node_modules', 'react-toolbox'),
            },

            ifDevClient({
              use: [
                'style-loader',
                {
                  loader: 'css-loader',
                  options: {
                    modules: true,
                    sourceMap: true,
                    importLoaders: 1,
                    localIdentName,
                  },
                },
                'postcss-loader',
              ],
            }),

            ifProdClient(() => ({
              use: [
                ...ExtractTextPlugin.extract({
                  fallback: 'style-loader',
                  use: [
                    `css-loader?modules=1&importLoaders=1&localIdentName=${localIdentName}`,
                  ],
                }),
              ],
            })),

            ifNode({
              use: [
                'classnames-loader',
                `css-loader/locals?modules=1&importLoaders=1&localIdentName=${localIdentName}`,
                'postcss-loader',
              ],
            }),
          ),
        ),
```

**Comment by unleashit** [comment]
Got it working by cloning react toolbox out of node_modules and into my project. It now works perfectly. So there's definitely something in the Webpack config that is excluding (or not including) the node_modules path from being processed on the server. The problem isn't the loader config from the repo that targets node_modules as I commented it out for testing. I also tried adding the React Toolbox path to JS/Babel processing in Values but that didn't do it either. 

It's not an ideal situation, but for now I'm stumped trying to get React Toolbox to work if it lives where it should be in node_modules. I'm sure I'm missing something obvious, but if anyone has a hint it would be appreciated.

**Comment by diondirza** [comment]
@unleashit I have made react-toolbox v2 working properly several months ago, now I have updated with latest master today. Maybe you could take a look at my branch [here](https://github.com/diondirza/react-universally/tree/feature/react-toolbox-sass)

**Comment by unleashit** [comment]
@diondirza actually your repo was a big help but it wasn't my loader configs. They aren't stock and all use explicit include/exclude arrays because I have a bunch of junk going on. It was as I suspected... something to do with the careful (in a good way) security measures of RU. After diffing your values.js, I could see you added /node_modules/ to  nodeExternalsFileTypeWhitelist.

As soon as I did that, the lights went on in the room! Thanks very much for your help.
